stdlib/Test/src/GenericDimensionfuls.jl

```diff
@@ -0,0 +1,109 @@
+Base.eps(::Type{GenericDimensionful{p,T}}) where {p,T<:AbstractFloat} = eps(T) # relative precision is dimensionless
+Base.eps(x::GenericDimensionful{p,T}) where {p,T<:AbstractFloat} = GenericDimensionful{p,T}(eps(x.val))
+Base.floatmin(::Type{GenericDimensionful{p,T}}) where {p,T<:AbstractFloat} = GenericDimensionful{p}(floatmin(T))
+Base.floatmin(::GenericDimensionful{p,T}) where {p,T<:AbstractFloat} = floatmin(GenericDimensionful{p,T})
+Base.floatmax(::Type{GenericDimensionful{p,T}}) where {p,T<:AbstractFloat} = GenericDimensionful{p}(floatmax(T))
+Base.floatmax(::GenericDimensionful{p,T}) where {p,T<:AbstractFloat} = floatmax(GenericDimensionful{p,T})
+
+# convert GenericDimensionful exponent p to a canonical form.  This
+# is not type stable, but it doesn't matter since it is used
+# at compile time (in generated functions), not runtime
+canonical_p(p) = isinteger(p) ? Int(p) : Rational{Int}(p)
+
+@generated Base.abs2(x::GenericDimensionful{p}) where {p} = :(GenericDimensionful{$(canonical_p(2p))}(abs2(x.val)))
+@generated Base.inv(x::GenericDimensionful{p}) where {p} = :(GenericDimensionful{$(canonical_p(-p))}(inv(x.val)))
+
+for f in (:isfinite, :isnan, :isreal, :isinf)
+    @eval Base.$f(x::GenericDimensionful) = $f(x.val)
+end
+for f in (:abs,:conj,:real,:imag,:complex,:+,:-)
+    @eval Base.$f(x::GenericDimensionful{p}) where {p} = GenericDimensionful{p}($f(x.val))
+end
+
+import Base: +, -, ==, !=, <, <=, isless, isequal, *, /, //, div, rem, mod, ^
+for op in (:+, :-)
+    @eval function $op(x::GenericDimensionful{p}, y::GenericDimensionful{p}) where {p}
+        v = $op(x.val, y.val)
+        GenericDimensionful{p}(v)
+    end
+end
+for op in (:(==), :(!=), :<, :<=, :isless, :isequal)
+    @eval $op(x::GenericDimensionful{p}, y::GenericDimensionful{p}) where {p} = $op(x.val, y.val)
+end
+# generated functions to allow type inference of the value of the exponent:
+for (f,op) in ((:_plus,:+),(:_minus,:-),(:_times,:*),(:_div,://))
+    @eval @generated function $f(v::T, ::GenericDimensionful{p}, ::Union{GenericDimensionful{q},Val{q}}) where {T,p,q}
+        s = $op(p, q)
+        :(GenericDimensionful{$(canonical_p(s)),$T}(v))
+    end
+end
+for (op,eop) in ((:*, :_plus), (:/, :_minus), (://, :_minus), (:div, :_minus))
+    @eval begin
+        $op(x::GenericDimensionful{p}, y::GenericDimensionful{q}) where {p,q} =
+            $eop($op(x.val, y.val),x,y)
+        $op(x::GenericDimensionful{p}, y::S) where {p,S<:Number} = $op(x,GenericDimensionful{0,S}(y))
+        $op(x::S, y::GenericDimensionful{p}) where {p,S<:Number} = $op(GenericDimensionful{0,S}(x),y)
+    end
+end
+# to fix an ambiguity
+//(x::GenericDimensionful, y::Complex) = x // GenericDimensionful{0,typeof(y)}(y)
+for op in (:rem, :mod)
+    @eval begin
+        $op(x::GenericDimensionful{p}, y::GenericDimensionful) where {p} = GenericDimensionful{p}($op(x.val, y.val))
+        $op(x::GenericDimensionful{p}, y::Number) where {p} = GenericDimensionful{p}($op(x.val, y))
```

Comment:
I don’t like defining `rem` and `mod` between values with different dimensions. To me, it seems more correct to throw a `DimensionMismatch` in that case. I realize that both Unitful.jl and the Dates stdlib allow it if the second argument is dimensionless, but what is `mod(GenericDimensionful{2}(5), GenericDimensionful{1}(3))` supposed to mean?

Reply:
We could define them.  `q = div(x, y)` an `r = rem(x, y)` are defined in general by `x = y*q + r`, which is unambiguous even when both `x` and `y` are dimensionful.

I agree that the code here is wrong when `y` is dimensionful, however.

Reply:
Is `DimensionMismatch` the right exception, by the way?  I thought that was only for mismatching `size`, not mismatched units…

Reply:
What I don’t like about it is that it is not invariant under unit conversion (e.g., `mod(1h, 5) != mod(60min, 5)`). I also think the result of `div` should always be an integer (it’s weird to me that `30min` is considered an integer but `0.5h` isn’t). But maybe I am too strict on this. There is some discussion on the topic here: https://github.com/PainterQubits/Unitful.jl/issues/350.

Regarding `DimensionMismatch`: In `Base` it is only used for `size`s, but Unitful uses it for mismatched unit dimensions as well.

Reply:
Fair enough; it wouldn't hurt to remove `div` and `rem` here since I don't think they are used in the tests.

Reply:
Actually, `div` and `rem` are used in range construction.

However, I think that in that specific case they are unambiguous: quotient and remainder for *two numbers with the same units*.   In that case, the results are unit-independent, so I left that case in and removed the case for differing units.